src/reactjs/useBaseQuery.ts

```diff
@@ -76,30 +72,21 @@ export function useBaseQuery<
       )
   )
 
-  let result = observer.getOptimisticResult(defaultedOptions)
+  const result = observer.getOptimisticResult(defaultedOptions)
 
-  React.useEffect(() => {
-    mountedRef.current = true
+  useSyncExternalStore(
```

Comment:
Is it intentional to not do anything with the result here? How does that work?

Reply:
it's a bit hacky. We need a stable result to subscribe to, which is `observer.getCurrentResult()`. However, the actual result displayed to the user is two lines above:

```
const result = observer.getOptimisticResult(defaultedOptions)
```

the optimistic result makes sure that we go into loading state immediately when mounting a component, even though it is not yet fetching. However, the result itself is not stored.

When we re-render because `getCurrentResult()` changes, we compute the optimistic result again, which will yield the same result, unless we are doing optimistic stuff.

Reply:
I see! So if I understand this correctly, this is safe because `getOptimisticResults()` only rely on things that we know will trigger re-renders elsewhere (e.g. via the `uSES` and `observer.getCurrentResult()`)?

Say `getOptimisticResults` would rely on some other piece of external state, that was not linked to `observer.getCurrentResult()`, in that case we would have a bug because that state could change without triggering a re-render of the component?

In a sense, we _could_ conceptually do something like this?

```js
const currentResult = useSyncExternalStore(
  React.useCallback(
    onStoreChange =>
      observer.subscribe(notifyManager.batchCalls(onStoreChange)),
    [observer]
  ),
  () => observer.getCurrentResult(),
  () => observer.getCurrentResult()
)

const optimisticResults = this.deriveOptimisticResults(currentResult)
```

Not a suggestion, just trying to formulate my understanding of why this hack is safe. 😄

Reply:
To be clear I'm trying to understand and explore 3 things:

* If and why this is safe (seems to be!)
* Potential avenues for this to _become_ unsafe via changes in other places of the code
* How optimistic results and hydration play together (because it seems conceptually like the dehydrated data that we plan to put on a context is a form of optimistic result)

Reply:
> In a sense, we could conceptually do something like this?

@Ephem yes, I think so. This is a good idea! From what I can see, the only thing that we need for optimisticResults is this:

https://github.com/tannerlinsley/react-query/blob/9d260e78b45e93ec6b5043de3bb61cac2ddb4345/src/core/queryObserver.ts#L417-L434

One problem might be that on the first render, `currentResult` is still `undefined` because we don't even have a cache entry yet. That's why `getOptimisticResult` primes the cache via `queryCache.build()`(should be safe, even in concurrent mode):

https://github.com/tannerlinsley/react-query/blob/9d260e78b45e93ec6b5043de3bb61cac2ddb4345/src/core/queryObserver.ts#L207-L219

so yeah, we might need to add `createResult()` into the mix if the uSES subscription returns `undefined`, but it still sounds like a better approach.

Reply:
ah, one thing that comes to mind is renders _not_ being triggered by react-query, but just top-down. For example, some local state might change, which you use inside the `select` function. Since `getOptimisticResult` calls `createResult`, which is also the function that executes select, we would also need to do that in the hypothetical `deriveOptimisticResults` 🤔

Reply:
for now, I would say to keep it as-is, and explore changes further down the road.